src/core/lib/event_engine/grpc_ares_wrapper.cc

```diff
@@ -0,0 +1,1019 @@
+  // Early out if the target is an ipv4 or ipv6 literal.
+  if (ResolveAsIPLiteralLocked()) {
+    return;
+  }
+  // TODO(yijiem): Early out if the target is localhost and we're on Windows.
+
+  // We add up pending_queries_ here since ares_gethostbyname may directly
+  // invoke the callback inline e.g. if there is any error with the input. The
+  // callback will invoke OnResolve with an error status and may start the
+  // shutdown process too early (before the second ares_gethostbyname) if we
+  // haven't added up here.
+  ++pending_queries_;
+  if (IsIpv6LoopbackAvailable()) {
+    ++pending_queries_;
+    auto* arg = new HostbynameArg();
+    arg->request = this;
+    arg->qtype = "AAAA";
+    ares_gethostbyname(channel_, host_.c_str(), AF_INET6,
+                       &GrpcAresHostnameRequest::OnHostbynameDoneLocked,
+                       static_cast<void*>(arg));
+  }
+  auto* arg = new HostbynameArg();
+  arg->request = this;
+  arg->qtype = "A";
+  ares_gethostbyname(channel_, host_.c_str(), AF_INET,
+                     &GrpcAresHostnameRequest::OnHostbynameDoneLocked,
+                     static_cast<void*>(arg));
+  // It's possible that ares_gethostbyname gets everything done inline.
+  if (!shutting_down_) {
+    Work();
+    StartTimers();
+  }
+}
+
+void GrpcAresHostnameRequest::OnResolveLocked(absl::StatusOr<Result> result) {
+  GPR_ASSERT(--pending_queries_ >= 0);
+  if (result_.ok()) {
+    if (result.ok()) {
+      result_->insert(result_->end(), result->begin(), result->end());
+    }
+    // else this error is ignored since we have already got a valid result.
+  } else {
+    if (result.ok()) {
+      // Got a valid result, store it.
+      result_ = std::move(result);
+    } else {
+      // Add result.status() as a child error of result_.
+      result_ = grpc_error_add_child(result_.status(), result.status());
+    }
+  }
+  if (pending_queries_ == 0) {
+    // Decrement the initial refcount. NOTE: this Unref will not trigger
+    // destruction. OnResolve is only called from c-ares callback which could be
+    // called from 3 places: 1. ares_process_fd in OnReadable/OnWritable
+    // where a RefCountedPtr is embedded in the closure; 2. ares_process_fd in
+    // OnAresBackupPollAlarm where a RefCountedPtr is embedded in the
+    // closure; 3. ares_gethostbyname in Start when OnResolve is called inline.
+    // We deliberately take a ref there through self.
+    Unref(DEBUG_LOCATION, "OnResolveLocked");
+    shutting_down_ = true;
+    if (cancelled_) {
+      // Cancel does not invoke on_resolve_.
+      return;
+    }
+    CancelTimers();
+    if (result_.ok() && !result_->empty()) {
+      SortResolvedAddressesLocked();
+    }
+    event_engine_->Run(
+        [on_resolve = std::move(on_resolve_), result = std::move(result_),
+         self = Ref(DEBUG_LOCATION, "OnResolveLocked")]() mutable {
+          grpc_core::ApplicationCallbackExecCtx callback_exec_ctx;
```

Comment:
This is a great callout!

I initially saw some tests failed because of `ExecCtx` was not on the stack in 2 cases:
1. When `on_resolve` gets called
2. When `on_resolve` is destroyed and the refcount inside `on_resolve` passed from the caller to the `DNSResolver` is destroyed (which triggers further destruction chain outward to a place that needs `ExecCtx`).

Both of these cases, the caller is `EventEngineClientChannelDNSResolver::EventEngineDNSRequestWrapper`, so I have added `ExecCtx` in the `on_resolve` callbacks and in the destructor of `EventEngineClientChannelDNSResolver::EventEngineDNSRequestWrapper`. This is so much convenient than trying to maintain this invariant inside EventEngine (assuming our tests cover all of the same scenarios). I have removed all `ExecCtx`s from EE's `grpc_ares_wrapper` and `PosixDNSResolver`. PTAL.